### PR TITLE
Revise Visualisation

### PR DESCRIPTION
**Pull request by gordonkoehn**
Rewriting the visualization of single mcmc run data. 

I am envisioning Snakemake rules for:
- [x] plot metrics to iteration, automatically Scale their axes and put labels on
- [x] Plot three most likely trees, with info of logprob and index 
- [x] plot initial tree 
- [x] plot log_prob to iteration
- [x] to make txt file which contains list of true tree samples found. 

- [x] Make an example experiment now, with all outputs

### CODE REVIEW

File: src/pyggdrasil/tree_inference/_config.py [hunk 1]
```diff
@@ -105,4 +105,4 @@ class McmcConfigOptions(Enum):
     """
 
     DEFAULT = (McmcConfig(),)
-    TEST = McmcConfig(n_samples=100)
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_config.py:108]: more uncertain posterior might be better for testing

File: src/pyggdrasil/visualize/_mcmc.py [hunk 2]
```diff
@@ -1,124 +1,122 @@
...
 from pathlib import Path
 
-from pyggdrasil import TreeNode
 from pyggdrasil.interface import PureMcmcData
-from pyggdrasil.distances import TreeSimilarityMeasure, calculate_distance_matrix
-from pyggdrasil.visualize import plot_tree_mcmc_sample, plot_tree_no_print
+from pyggdrasil.visualize import plot_tree_mcmc_sample
+
+# use SVG backend for matplotlib
+matplotlib.use("SVG")
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/_mcmc.py:13]: Choosing a non-interactive (allows for multitheading), system-independent backend for plot generation.

File: src/pyggdrasil/visualize/_mcmc.py [hunk 3]
```diff
@@ -1,124 +1,122 @@
 """Methods for visualizing mcmc samples."""
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/_mcmc.py:1]: As expected, I largely rewrote this entire file. - Learned how to modularize plotting functions in the last PR though. - Was still very useful.

File: src/pyggdrasil/visualize/_mcmc.py [hunk 4]
```diff
@@ -138,72 +136,89 @@ def _save_top_trees_plots(data: PureMcmcData, output_dir: Path) -> None:
         plot_tree_mcmc_sample(each, output_dir, save_name=f"top_tree_{rank}")
         rank += 1
 
-
-def make_mcmc_run_panel(
-    data: PureMcmcData,
-    true_tree: TreeNode,
-    similarity_measure: TreeSimilarityMeasure,
-    out_dir: Path,
-) -> None:
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/_mcmc.py:147]: Abandon - rather have single-responsibility functions and generate all plots of interest with a snakemake rule.

File: src/pyggdrasil/visualize/_tree.py [hunk 5]
```diff
@@ -88,7 +92,7 @@ def plot_tree(
 
     # Calculate an appropriate figure size
     node_width = 0.5  # Width of each node in the figure
-    node_height = 0.75  # Height of each node in the figure
+    node_height = 1.3  # Height of each node in the figure
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/_tree.py:95]: adjusting node height - top and bottom nodes were cut off. Apparently, that's nothing Matplotlib can to automatically.

File: src/pyggdrasil/visualize/_tree.py [hunk 6]
```diff
@@ -166,9 +170,13 @@ def plot_tree(
...
     plt.text(1, 0.97, description, dict(size=15), transform=ax1.transAxes, va="top")
 
-    plt.savefig(fullpath + ".svg", bbox_inches="tight")
+    # check if fullpath already has .svg extension, else add it
+    if not fullpath.endswith(".svg"):
+        fullpath += ".svg"
+    # save
+    plt.savefig(fullpath, bbox_inches="tight", format="svg")
     plt.close()
-    logger.info(f"Saved tree plot to {fullpath}.svg")
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/_tree.py:171]: different functions call it differently unfortunately - might fix later

File: workflows/Snakefile [hunk 7]
```diff
@@ -1,321 +1,13 @@
...
 
-import pyggdrasil as yg
+################################################################################
+# Imports Utilities
+include: "tree_inference.smk"
+include: "analyze.smk"
+include: "visualize.smk"
+# Import Experiments
+include: "mark00.smk"
+################################################################################
```
Comment (gordonkoehn) [inline workflows/Snakefile:11]: Structured the snakemake workflows now alike our modules and experiments. The main snakemake is just to bind them all together.

File: workflows/mark00.smk [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+true_tree_seed = 5
+true_tree_n_nodes = nodes
+# Initial Tree
+initial_tree_type = TreeType("d")
+initial_tree_seed = 4545
+initial_tree_n_nodes = nodes
+# MCMC Parameters
+mcmc_seed = 42
+mcmc_config_id = McmcConfig(n_samples=100, fpr=0.4, fnr=0.4).id()
+###############################################
```
Comment (gordonkoehn) [inline workflows/mark00.smk:31]: One place to specify one MCMC chain run. - Will think of the best way to run multiple chains for the experiments.

File: workflows/tree_inference.smk [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        """
+        poetry run python {input.script} \
+            --out_dir ../data/{wildcards.experiment}/trees \
+            --seed {wildcards.tree_seed} \
+            --n_nodes {wildcards.n_nodes} \
+            --tree_type {wildcards.tree_type}
+        """
+
+rule make_star_tree:
+    """"Make a tree (TreeNode) of star topology and save it as JSON."""
```
Comment (gordonkoehn) [inline workflows/tree_inference.smk:27]: separate rule as a star tree needs no seed. - Hence output string is shorter. Trouble with string matching :(

File: workflows/tree_inference.smk [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        --out_dir ../data/{wildcards.experiment}/mutations \
+        """
+
+
+# choose the optimal move probabilities, by enums
+#optimal_MP_conf = MoveProbConfigOptions.OPTIMAL
+
+
+# Set the Model manually
+#customMcmcConfig = McmcConfigOptions.TEST.value
```
Comment (gordonkoehn) [inline workflows/tree_inference.smk:68]: note - may remove

**Review by gordonkoehn — COMMENTED** [review]
Fully functional single MCMC run specification, analysis and visualization.

File: workflows/Snakefile (hunk 7 above)
Comment (pawel-czyz) [reply workflows/Snakefile:11]: Very, very nice!

**Review by pawel-czyz — APPROVED** [review]
Nice! I think the structure is much more readable for me now. I have problems with understanding string matching (it is more advanced than what I have been doing so far), but I'm glad it works. Very good work, please merge :slightly_smiling_face:

File: workflows/visualize.smk [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        init_tree_id = "T.*",
+    output:
+        plot = '../data/{experiment}/plots/MCMC_{mcmc_seed,\d+}-{mutation_data_id}-i{init_tree_id}-{mcmc_config_id}/init_tree.svg',
+    run:
+        in_fp = Path(input.initial_tree)
+        out_fp = Path(output.plot)
+        # read in tree
+        initial_tree = yg.serialize.read_tree_node(in_fp)
+        # plot the tree
+        yg.visualize.plot_tree_no_print(initial_tree, save_name=out_fp.name.__str__(), save_dir=out_fp.parent)
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply workflows/visualize.smk:52]: ok noted - thank you.

File: src/pyggdrasil/visualize/_mcmc.py [hunk 12]
```diff
@@ -1,124 +1,122 @@
...
+) -> None:
     """Save plot of log probability vs iteration number to disk.
 
     Args:
-        data: PureMcmcData
-            Data from MCMC runs.
         output_dir: Path
             Output directory to save the plot.
+        iterations: list[int]
+            Iteration numbers.
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/visualize/_mcmc.py]: 1-indexed. Added a note to the docstring.

File: src/pyggdrasil/visualize/_tree.py [hunk 13]
```diff
@@ -166,9 +170,13 @@ def plot_tree(
 
     plt.text(1, 0.97, description, dict(size=15), transform=ax1.transAxes, va="top")
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/visualize/_tree.py:171]: yes, I see what you mean. Might change if that function needs some work later. Appreciate all suggestions!

**Pull request merged**